expression/builtin.go

```diff
@@ -221,76 +219,83 @@ func (b *baseBuiltinFunc) getArgs() []Expression {
 	return b.args
 }
 
-func (b *baseBuiltinFunc) evalInt(row []types.Datum) (int64, bool, error) {
-	val, err := b.self.eval(row)
-	if err != nil || val.IsNull() {
-		return 0, val.IsNull(), errors.Trace(err)
+// eval should only be called in test files, and it should be removed after all tests being rewritten.
+func (b *baseBuiltinFunc) eval(row []types.Datum) (d types.Datum, err error) {
+	var (
+		res    interface{}
+		isNull bool
+	)
+	switch fieldTp2EvalTp(b.tp) {
+	case tpInt:
+		var intRes int64
+		intRes, isNull, err = b.self.evalInt(row)
+		if mysql.HasUnsignedFlag(b.tp.Flag) {
+			res = uint64(intRes)
+		} else {
+			res = intRes
+		}
+	case tpReal:
+		res, isNull, err = b.self.evalReal(row)
+	case tpDecimal:
+		res, isNull, err = b.self.evalDecimal(row)
+	case tpDatetime, tpTimestamp:
+		res, isNull, err = b.self.evalTime(row)
+	case tpDuration:
+		res, isNull, err = b.self.evalDuration(row)
+	case tpJSON:
+		res, isNull, err = b.self.evalJSON(row)
+	case tpString:
+		res, isNull, err = b.self.evalString(row)
```

Comment:
default?

Reply:
Yes, mainly copied from `ScalarFunction.Eval()`. This function now only used in unit tests, we should remove this API in the future.